### PR TITLE
Feature/jshell

### PR DESCRIPTION
**Pull request by Alathreon**
* JShell feature added

* Added option to get the snippets of another user

* Refactored JShell eval into its own class

* Changed RateLimiter so it's a global limiter and not a per user limiter + merged user and oneOffSession together

* Added context action for running java code

* Fixing sonar wranings for jshell

## Config

The config was changed, the following has been introduced:

```json
"jshell": {
    "baseUrl": "<put_jshell_rest_api_url_here>",
    "rateLimitWindowSeconds": 10,
    "rateLimitRequestsInWindow": 3
}
```

### CODE REVIEW

File: application/config.json.template [hunk 1]
```diff
@@ -90,4 +90,9 @@
     "logErrorChannelWebhook": "<put_your_webhook_here>",
     "openaiApiKey": "<check pins in #tjbot_discussion for the key>",
     "sourceCodeBaseUrl": "<https://github.com/<your_account_here>/<your_repo_here>/blob/master/application/src/main/java/>"
+    "jshell": {
+        "baseUrl": "http://localhost:8080/jshell/",
```
Comment (Zabuzard) [inline application/config.json.template:94]: its better to leave it out entirely if we cant have a meaningful default.
so `<put_your_api_endpoint_here>`

File: application/src/main/java/org/togetherjava/tjbot/config/Config.java [hunk 2]
```diff
@@ -330,4 +333,13 @@ public String getOpenaiApiKey() {
     public String getSourceCodeBaseUrl() {
         return sourceCodeBaseUrl;
     }
+
+    /**
+     * The configuration about jshell.
+     * 
+     * @return the jshell configuration
+     */
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/config/Config.java:341]: worthless javadoc. please improve it

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    private static final int MIN_MESSAGE_INPUT_LENGTH = 0;
+    private static final int MAX_MESSAGE_INPUT_LENGTH = TextInput.MAX_VALUE_LENGTH;
+
+    private final JShellEval jshellEval;
+
+    /**
+     * Creates an instance of the command.
+     */
+    public JShellCommand(JShellEval jshellEval) {
+        super(JSHELL_COMMAND, "JShell as a command.", CommandVisibility.GUILD);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:53]: bad description. users already know its a command when reading this text. provide explanation what jshell is instead

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+        this.jshellEval = jshellEval;
+
+        getData().addSubcommands(
+                new SubcommandData(JSHELL_VERSION_SUBCOMMAND, "Get the version of JShell"),
+                new SubcommandData(JSHELL_EVAL_SUBCOMMAND,
+                        "Evaluate java code in JShell, don't fill the optional parameter to access a bigger input box.")
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: maybe reword to "... submit the command without code for inputting longer, multi-line code"

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                            .addOption(OptionType.STRING, JSHELL_CODE_PARAMETER,
+                                    "Code to evaluate. If not supplied, open an inout box.")
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:62]: typo "inout box", but id reword it to maybe "... Leave empty to input longer, multi-line code"

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                            .addOption(OptionType.BOOLEAN, JSHELL_STARTUP_SCRIPT_PARAMETER,
+                                    "If the startup script should be loaded, true by default."),
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:64]: the description is the place to explain what that startup script does, why would the user want to load it? what does it do? that has to be here

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+                new SubcommandData(JSHELL_SNIPPETS_SUBCOMMAND,
+                        "Get the evaluated snippets of the user who sent the command, or the user specified user if any.")
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:66]: u wrote the description facing the programmer. but the description is supposed to face the user. i.e. `you` is the user, not the programmer.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                            .addOption(OptionType.USER, JSHELL_USER_PARAMETER,
+                                    "User to get the snippets from. If null, get the snippets of the user who sent the command.")
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:68]: same here. its `your snippets`

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+                            .addOption(OptionType.BOOLEAN, JSHELL_INCLUDE_STARTUP_SCRIPT_PARAMETER,
+                                    "if the startup script should be included, false by default."),
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:70]: confusing, there was already a parameter like this?

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+    }
+
+    private void handleVersionCommand(SlashCommandInteractionEvent event) {
+        String code = """
+                System.out.println("```");
+                System.out.println("Version: " + Runtime.version());
+                System.out.println("Vendor:  " + System.getProperty("java.vendor"));
+                System.out.println("OS:      " + System.getProperty("os.name"));
+                System.out.println("Arch:    " + System.getProperty("os.arch"));
+                 System.out.println("```");""";
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:109]: the last line has an extra space, looks odd

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+            }
+
+            if (snippets.stream().noneMatch(s -> s.length() >= MessageEmbed.VALUE_MAX_LENGTH)
+                    && snippets.stream()
+                        .mapToInt(s -> (s + "Snippet 10```java\n```").length())
+                        .sum() < MessageEmbed.EMBED_MAX_LENGTH_BOT - 100
+                    && snippets.size() <= 25/*
+                                             * Max visible embed fields in an embed TODO replace
+                                             * with constant
+                                             */) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:191]: replace with well named constant

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandler;
+import java.net.http.HttpResponse.BodyHandlers;
+
+public class JShellApi {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java:22]: this class needs some javadoc

File: application/src/main/java/org/togetherjava/tjbot/features/utils/RateLimiter.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Custom rate limiter.
+ */
+public class RateLimiter {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/RateLimiter.java:12]: worthless javaodc. please improve. especially because this is a general purpose utility (judging by the package)

File: application/src/main/java/org/togetherjava/tjbot/features/utils/RequestFailedException.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
+package org.togetherjava.tjbot.features.utils;
+
+public class RequestFailedException extends Exception {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/utils/RequestFailedException.java:3]: i dont quite see the point of introducing a checked exception and an unchecked variant at the same time. just go with the unchecked

**Review by Zabuzard — CHANGES_REQUESTED** [review]

**Comment by Zabuzard** [comment]
please fix the sonar warnings as well, thanks

File: application/config.json.template (hunk 1 above)
Comment (Alathreon) [reply application/config.json.template:94]: Since the api should be on the same machine, and should have port 8080 by default, it totally makes sense to have this default

File: application/src/main/java/org/togetherjava/tjbot/features/utils/RequestFailedException.java (hunk 14 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/RequestFailedException.java:3]: I need to use an unchecked exception in JShellAPI because it doesn't support checked exceptions (more specifically ResponseUtils.ofJson), the unchecked exception is purely internal and not leaked outside of JShellAPI.
I use a checked exception when using JShellAPI but they must be handled, and I already forgotten several times during the coding of the commands to handle them, so it's too important to make it unchecked.

File: application/src/main/java/org/togetherjava/tjbot/config/Config.java (hunk 2 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/config/Config.java:341]: I added it to be consistant, since all other getter have javadoc even if it is worthless, do I still remove it ? I don't know what I can really add to it.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 9 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:70]: It's different, it's if the user wants to also display the startupt script in the snippets, I don't know if I should just delete this option.

File: application/src/main/java/org/togetherjava/tjbot/features/code/EvalCodeCommand.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import net.dv8tion.jda.api.EmbedBuilder;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+
+import org.togetherjava.tjbot.features.jshell.JShellEval;
+import org.togetherjava.tjbot.features.jshell.render.Colors;
+import org.togetherjava.tjbot.features.utils.CodeFence;
+import org.togetherjava.tjbot.features.utils.RequestFailedException;
+
+/**
+ * Evaluates the given code.
```
[the remark this answers is not in the document]
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/code/EvalCodeCommand.java]: Improved, tell me what you think.

File: application/config.json.template (hunk 1 above)
Comment (Zabuzard) [reply application/config.json.template:94]: the config template is not for the production machine. it is for the developer. and the developer does not run a second program when developing tj-bot. in fact, there is a high risk that some other thing is currently running on their machine on 8080.

if you can not provide an external link that is available out-the-box when a dev runs tj-bot locally, u shouldnt put sth meaningful there and the feature should be disabled. (similar to for example the discord log-forwarder)

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+    private static final String JSHELL_COMMAND = "jshell";
+    private static final String JSHELL_VERSION_SUBCOMMAND = "version";
+    private static final String JSHELL_EVAL_SUBCOMMAND = "eval";
+    private static final String JSHELL_SNIPPETS_SUBCOMMAND = "snippets";
+    private static final String JSHELL_CLOSE_SUBCOMMAND = "shutdown";
+    private static final String JSHELL_STARTUP_SCRIPT_SUBCOMMAND = "startup-script";
+    private static final String JSHELL_CODE_PARAMETER = "code";
+    private static final String JSHELL_STARTUP_SCRIPT_PARAMETER = "startup-script";
+    private static final String JSHELL_USER_PARAMETER = "user";
+    private static final String JSHELL_INCLUDE_STARTUP_SCRIPT_PARAMETER = "include-startup-script";
```
[the remark this answers is not in the document]
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Improved, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 3 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Improved, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 4 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Fixed.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 5 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Fixed.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 7 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Improved, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 11 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Fixed.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java (hunk 12 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java:34]: Improved, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/utils/RateLimiter.java (hunk 13 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/RateLimiter.java]: Improved, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/config/Config.java (hunk 2 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/config/Config.java:341]: the others are not worthless, they add extra information - or they are bad as well.

u cant remove it, since its public. please put in some effort and describe what "jshell" is and does in the context of tj-bot. the config is the place where people will read about it the first time and they have to understand what this config is about. i.e. that its an external api that executes java code using JShell.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 9 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:78]: u have to rename and reword it, so that its clear and not confusing anymore - then its okay

File: application/src/main/java/org/togetherjava/tjbot/features/utils/RequestFailedException.java (hunk 14 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/utils/RequestFailedException.java:3]: mh, fair

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 10 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Fixed.

File: application/src/main/java/org/togetherjava/tjbot/features/code/CodeMessageHandler.java [hunk 17]
```diff
@@ -63,6 +63,8 @@ public final class CodeMessageHandler extends MessageReceiverAdapter implements
 
     /**
      * Creates a new instance.
+     * 
+     * @param jshellEval the jshell evaluation instance used in the code actions
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/code/CodeMessageHandler.java:67]: this style of javadoc is useless fluff. ur just describing in words and sentences what is known already. javadoc like that has no substance, it is not helpful.

u might argue "yes, i didnt want to put it, but im forced to", but the point of the rule is to learn how to write meaningful javadoc.

this is more of a general critique on the whole javadoc in this PR (although u already improved a lot of it).

for this particular one, maybe describe what it is used for. `@param jshellEval used to execute Java code and show the user the output`.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 18]
```diff
@@ -20,26 +20,30 @@
...
+import org.togetherjava.tjbot.features.utils.MessageUtils;
 import org.togetherjava.tjbot.features.utils.RequestFailedException;
 
 import javax.annotation.Nullable;
 
 import java.util.List;
 import java.util.Objects;
 
+/**
+ * The JShell command, provide functionalities to create JShell sessions, evaluate code, etc.
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:33]: give a quick and simple usage example (see other slash command javadocs)

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 19]
```diff
@@ -48,51 +52,53 @@ public class JShellCommand extends SlashCommandAdapter {
 
     /**
      * Creates an instance of the command.
+     * 
+     * @param jshellEval the jshell evaluation instance used in the jshell command
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:56]: (same with the "useless javadoc")

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 20]
```diff
@@ -48,51 +52,53 @@ public class JShellCommand extends SlashCommandAdapter {
...
      */
     public JShellCommand(JShellEval jshellEval) {
-        super(JSHELL_COMMAND, "JShell as a command.", CommandVisibility.GUILD);
+        super(JSHELL_COMMAND,
+                "JShell as a command, a Read–eval–print loop which allows to rapidly and simply execute java code.",
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: "read-eval-print loop" sounds too technical imo. hard to understand as a noob. also, its kinda clear that its a command when the user reads this, as they already typed `/jshell` at that point. maybe
`Executes the given Java code and shows the output.`

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java [hunk 21]
```diff
@@ -12,10 +12,23 @@
...
+/**
+ * Allows to render JShell results.
+ */
+class ResultRenderer {
 
+    /**
+     * Renders a JShell result to an embed.
+     * 
+     * @param originator the user from who to display snippet ownership
+     * @param originalCode the original code to display
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java:26]: since nullable, what happens if null? should be mentioned in the doc

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java [hunk 22]
```diff
@@ -19,21 +19,51 @@
 import java.net.http.HttpResponse.BodyHandler;
 import java.net.http.HttpResponse.BodyHandlers;
 
+/**
+ * Allows to interact with the unofficial JShell REST API of the Together-Java JShell backend
+ * project.
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java:24]: might just drop a link to it as well. `<a href="...">JShell REST API</a>`

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/dto/JShellResult.java [hunk 23]
```diff
@@ -1,11 +1,39 @@
...
+ * 
+ * @param status status of the snippet
+ * @param type type of the snippet
+ * @param id id of the snippet
+ * @param source source code of the snippet
+ * @param result result of the snippet, nullable
+ * @param exception thrown exception, nullable
+ * @param stdoutOverflow if stdout has overflowed and was truncated
+ * @param stdout what was printed by the snippet
+ * @param errors the compilations errors of the snippet
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/dto/JShellResult.java:18]: "useless javadoc". leaves any reader still puzzled what what means. what is a snippet status? what is a snippet type? what happens if result is null? what if exception is null? who throws the exception? what does it indicate? ...

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+                if (e.getStatus() == JShellApi.SESSION_NOT_FOUND) {
+                    interactionHook.editOriginalEmbeds(createSessionNotFoundErrorEmbed(user))
+                        .queue();
+                } else {
+                    interactionHook.editOriginalEmbeds(createUnexpectedErrorEmbed(user, e)).queue();
+                }
+                return;
+            }
+
+            if (snippets.stream().noneMatch(s -> s.length() >= MessageEmbed.VALUE_MAX_LENGTH)
```
Comment (SquidXTV) [inline application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:188]: why not splitting the logic into private methods?
so the `if` doesn't  will not be this clumped up

**Review by SquidXTV — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java (hunk 22 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java:24]: I thought of that, but since the name of the API isn't fixed and might change, I am afraid that we forget to change this link.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java (hunk 21 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java:26]: I thought it was obvious enough, this method render stuff, so if it's null it doesn't render it.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 20 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: "JShell integration in discord, allows to rapidly and simply execute java code" ?

File: application/src/main/java/org/togetherjava/tjbot/config/JShellConfig.java [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+public record JShellConfig(String baseUrl, int rateLimitWindowSeconds,
+        int rateLimitRequestsInWindow) {
+    /**
+     * Creates a JShell config.
+     * 
+     * @param baseUrl the base url of the JShell REST API, must be valid
+     * @param rateLimitWindowSeconds the number of seconds of the {@link RateLimiter rate limiter}
+     *        for jshell commands and code actions, must be higher than 0
+     * @param rateLimitRequestsInWindow the number of requests of the {@link RateLimiter rate
+     *        limiter} for jshell commands and code actions, must be higher than 0
```
[the remark this answers is not in the document]
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/config/JShellConfig.java:27]: I don't believe it's needed

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java (hunk 21 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java:26]: not obvious to me, please put it, thanks

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 20 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: Since this is a user facing explanation, id prefer a text that is less general and "buy this product, its good".
Maybe "Executes Java code using JShell, shows the output"

File: application/src/main/java/org/togetherjava/tjbot/config/JShellConfig.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/config/JShellConfig.java:27]: `throws` is usually only added if u want to explain sth about it (or if its a checked exception, since then its necessary to explain).
but `IllegalArgumentException` when the arguments are illegal is just classic validation stuff - id consider it idiomatic enough to go without explanation

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 20 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: This would be the description of /jshell eval, not the whole /jshell 🤔

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: yeah. but it is its main functionality and it will be the first thing users read when they explore the command in discord by typing `/` 🤷

Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: I think I'll add a /jshell help instead, and keep the generic explanation, what do you think ?

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 6 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:72]: Is it resolved ?

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 20 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java:60]: i dont think its needed. i doubt anyone would use it. i just feel like ur explanation is sorta missing the information the a regular user in our server would need to read.

File: application/src/main/java/org/togetherjava/tjbot/features/code/CodeMessageHandler.java (hunk 17 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/code/CodeMessageHandler.java]: Changed, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 19 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Changed, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 20 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Changed, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java (hunk 21 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/ResultRenderer.java]: Changed, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java (hunk 22 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/JShellApi.java]: Changed, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/dto/JShellResult.java (hunk 23 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/backend/dto/JShellResult.java]: >  what is a snippet status? what is a snippet type?
This is explained in the corresponding classes, check them.

For the rest, I changed things, tell me what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java (hunk 24 above)
Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: Changed, tell me what you think.

Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: I didnt really mean the if itself, I meant to have a `private boolean ...` method for `snippets.stream().noneMatch(s -> s.length() >= MessageEmbed.VALUE_MAX_LENGTH)`, `snippets.stream().mapToInt(s -> (s + "Snippet 10```java\n```").length()).sum() < MessageEmbed.EMBED_MAX_LENGTH_BOT - 100`, ...

Comment (Alathreon) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: How is it now ?

Comment (SquidXTV) [reply application/src/main/java/org/togetherjava/tjbot/features/jshell/JShellCommand.java]: yeah I am fine with that

**Review by SquidXTV — APPROVED** [review]

**Review by Zabuzard — APPROVED** [review]
